src/tests.rs

```diff
@@ -0,0 +1,258 @@
+    struct S {
+        field_1: u32,
+        field_2: bool,
+    }
+
+    let mut val = S {
+        field_1: 60,
+        field_2: true,
+    };
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(&mut val)) };
+    unsafe {
+        volatile.map_mut(|s| NonNull::new(core::ptr::addr_of_mut!((*s.as_ptr()).field_1)).unwrap())
+    }
+    .update(|v| *v += 1);
+    let mut field_2 = unsafe {
+        volatile.map_mut(|s| NonNull::new(core::ptr::addr_of_mut!((*s.as_ptr()).field_2)).unwrap())
+    };
+    assert!(field_2.read());
+    field_2.write(false);
+    assert_eq!(
+        val,
+        S {
+            field_1: 61,
+            field_2: false
+        }
+    );
+}
+
+#[test]
+fn test_struct_macro() {
+    #[derive(Debug, PartialEq)]
+    struct S {
+        field_1: u32,
+        field_2: bool,
+    }
+
+    let mut val = S {
+        field_1: 60,
+        field_2: true,
+    };
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(&mut val)) };
+    let mut field_1 = map_field_mut!(volatile.field_1);
+    field_1.update(|v| *v += 1);
+    let mut field_2 = map_field_mut!(volatile.field_2);
+    assert!(field_2.read());
+    field_2.write(false);
+    assert_eq!(
+        val,
+        S {
+            field_1: 61,
+            field_2: false
+        }
+    );
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+fn test_slice() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    volatile.index_mut(0).update(|v| *v += 1);
+
+    let mut dst = [0; 3];
+    volatile.copy_into_slice(&mut dst);
+    assert_eq!(dst, [2, 2, 3]);
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+#[should_panic]
+fn test_bounds_check_1() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    volatile.index_mut(3);
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+#[should_panic]
+fn test_bounds_check_2() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    volatile.index_mut(2..1);
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+#[should_panic]
+fn test_bounds_check_3() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    volatile.index_mut(4..); // `3..` is is still ok (see next test)
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+fn test_bounds_check_4() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    assert_eq!(volatile.index_mut(3..).len(), 0);
+}
+
+#[cfg(feature = "unstable")]
+#[test]
+#[should_panic]
+fn test_bounds_check_5() {
+    let val: &mut [u32] = &mut [1, 2, 3];
+    let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(val)) };
+    volatile.index_mut(..4);
+}
+
+#[cfg(feature = "unstable")]
```

Comment:
Are there plans to have tests for the `"very_unstable"` features?" Maybe just doc-tests are enough?